src/container.cpp

```diff
@@ -446,63 +444,63 @@ Cylinder* Container::queryDestination(int32_t& index, const Thing& thing, Item**
 		the slot position as if the bag has 20 slots
 		*/
 		index = INDEX_WHEREEVER;
-		*destItem = nullptr;
+		destItem = nullptr;
 	}
 
-	const Item* item = thing.getItem();
+	auto item = thing->getItem();
 	if (!item) {
-		return this;
+		return shared_from_this();
 	}
 
 	if (index != INDEX_WHEREEVER) {
-		Item* itemFromIndex = getItemByIndex(index);
+		auto itemFromIndex = getItemByIndex(index);
 		if (itemFromIndex) {
-			*destItem = itemFromIndex;
+			destItem = itemFromIndex;
 		}
 
-		Cylinder* subCylinder = dynamic_cast<Cylinder*>(*destItem);
-		if (subCylinder) {
+		if (destItem) {
+			auto subCylinder = destItem;
 			index = INDEX_WHEREEVER;
-			*destItem = nullptr;
+			destItem = nullptr;
 			return subCylinder;
 		}
 	}
 
 	bool autoStack = !hasBitSet(FLAG_IGNOREAUTOSTACK, flags);
-	if (autoStack && item->isStackable() && item->getParent() != this) {
-		if (*destItem && (*destItem)->equals(item) && (*destItem)->getItemCount() < ITEM_STACK_SIZE) {
-			return this;
+	if (autoStack && item->isStackable() && item->getParent().get() != this) {
+		if (destItem && *destItem == *item && destItem->getItemCount() < ITEM_STACK_SIZE) {
+			return shared_from_this();
 		}
 
 		// try find a suitable item to stack with
 		uint32_t n = 0;
-		for (Item* listItem : itemlist) {
-			if (listItem != item && listItem->equals(item) && listItem->getItemCount() < ITEM_STACK_SIZE) {
-				*destItem = listItem;
+		for (const auto& listItem : itemList) {
+			if (listItem != item && *listItem == *item && listItem->getItemCount() < ITEM_STACK_SIZE) {
+				destItem = listItem;
 				index = n;
-				return this;
+				return shared_from_this();
 			}
 			++n;
 		}
 	}
-	return this;
+	return shared_from_this();
 }
 
-void Container::addThing(Thing* thing) { return addThing(0, thing); }
+void Container::addThing(std::shared_ptr<Thing> thing) { return addThing(0, thing); }
 
-void Container::addThing(int32_t index, Thing* thing)
+void Container::addThing(int32_t index, std::shared_ptr<Thing> thing)
 {
 	if (index >= static_cast<int32_t>(capacity())) {
 		return /*RETURNVALUE_NOTPOSSIBLE*/;
 	}
 
-	Item* item = thing->getItem();
+	auto item = thing->getItem();
```

Comment:
Could this be `const auto`
```suggestion
	const auto item = thing->getItem();
```